lib/iris/pandas.py

```diff
@@ -116,6 +205,290 @@ def as_cube(pandas_array, copy=True, calendars=None):
+    >>> my_df = my_df.sort_index()
+    >>> converted_cube = as_cubes(my_df)[0]
+    >>> print(converted_cube)
+    air_temperature / (unknown)         (latitude: 2; longitude: 2)
+        Dimension coordinates:
+            latitude                             x             -
+            longitude                            -             x
+
+    """
+    calendars = calendars or {}
+    aux_coord_cols = aux_coord_cols or []
+    cell_measure_cols = cell_measure_cols or []
+    ancillary_variable_cols = ancillary_variable_cols or []
+
+    is_series = isinstance(pandas_array, pandas.Series)
+
+    if copy:
+        pandas_array = pandas_array.copy()
+
+    pandas_index = pandas_array.index
+    if not pandas_index.is_unique:
+        message = (
+            f"DataFrame index ({pandas_index.names}) is not unique per "
+            "row; cannot be used for DimCoords."
+        )
+        raise ValueError(message)
+
+    if not pandas_index.is_monotonic:
+        # Need monotonic index for use in DimCoord(s).
+        # This function doesn't sort_index itself since that breaks the
+        #  option to return a data view instead of a copy.
+        message = (
+            "Pandas index is not monotonic. Consider using the "
+            "sort_index() method before passing in."
+        )
+        raise ValueError(message)
+
+    cube_shape = getattr(pandas_index, "levshape", (pandas_index.nunique(),))
+
+    cube_kwargs = {}
+
+    def format_dimensional_metadata(dm_class_, values_, name_, dimensions_):
+        # Common convenience for to get the right DM in the right format for
+        #  Cube creation.
+        calendar = calendars.get(name_)
+        instance = _get_dimensional_metadata(
+            name_, values_, calendar, dm_class_
+        )
+        return (instance, dimensions_)
+
+    # DimCoords.
+    dim_coord_kwarg = []
+    for ix, dim_name in enumerate(pandas_index.names):
+        if hasattr(pandas_index, "levels"):
+            coord_points = pandas_index.levels[ix]
+        else:
+            coord_points = pandas_index
+        new_dim_coord = format_dimensional_metadata(
+            DimCoord, coord_points, dim_name, ix
+        )
+        dim_coord_kwarg.append(new_dim_coord)
+    cube_kwargs["dim_coords_and_dims"] = dim_coord_kwarg
+
+    # Other dimensional metadata.
+    class_arg_mapping = [
+        (AuxCoord, aux_coord_cols, "aux_coords_and_dims"),
+        (CellMeasure, cell_measure_cols, "cell_measures_and_dims"),
+        (
+            AncillaryVariable,
+            ancillary_variable_cols,
+            "ancillary_variables_and_dims",
+        ),
+    ]
+
+    non_data_names = []
+    for dm_class, column_names, kwarg in class_arg_mapping:
+        class_kwarg = []
+        if is_series and column_names:
+            message = (
+                "The input pandas_array is a Series; ignoring "
+                f"{dm_class.__name__} columns: {column_names} ."
+            )
+            warnings.warn(message)
```

Comment:
Trouble with this is that the warning could be displayed up to 3 times. I'm always a bit wary of adding warnings after having heard a fair amount of complaints about them from our users. Could you rather put this if block outside the for loop and instead aggregate the checks into one warning?

Reply:
7e8ff414d81e77d52a12d0ff1de5e997bacb6c7f